### PR TITLE
copilot: Add support for new models

### LINKED ISSUE #19963: Add other LLMs via copilot (sonnet 3.5, gemini, o1-mini/preview)
### Check for existing issues

- [X] Completed

### Describe the feature

Github just announced they're including other models in copilot, it'd be great to use them through zed

### If applicable, add mockups / screenshots to help present your vision of the feature

_No response_

### PR DESCRIPTION
**Pull request by overjt**
Closes #19963

This PR implements integration with the newly announced GitHub Copilot LLM models, including:
- Claude 3.5 Sonnet
- o1-mini
- o1-preview

Release Notes:

- N/A

### CODE REVIEW

**Comment by notpeter** [comment]
Hi @overjt, I tested this branch. It didn't work for either of the o1 models.  The request sent but no assistant response after ~5secs (no error in the logs, no timeout message).  I wonder whether we need to add support for longer timeouts and perhaps and `max_completion_tokens` as here:
- https://github.com/zed-industries/zed/pull/17796
- https://github.com/zed-industries/zed/pull/17804

Does Claude work for you? My Copilot account does not yet have Claude enabled so I got this instead (not unexpected):
Screenshot 2024-10-31 at 7 47 34

**Comment by SirSilver** [comment]
> Hi @overjt, I tested this branch. It didn't work for either of the o1 models. The request sent but no assistant response after ~5secs (no error in the logs, no timeout message). I wonder whether we need to add support for longer timeouts and perhaps and `max_completion_tokens` as here:
> 
> * [OpenAI o1-preview and o1-mini support #17796](https://github.com/zed-industries/zed/pull/17796)
> * [Add ability to use o1-preview and o1-mini as custom models #17804](https://github.com/zed-industries/zed/pull/17804)
> 
> Does Claude work for you? My Copilot account does not yet have Claude enabled so I got this instead (not unexpected): Screenshot 2024-10-31 at 7 47 34

Same with o1 models.
Clade works for me.

**Comment by notpeter** [comment]
Would you be willing to split this PR so we could merge the working Anthropic models and then ship o1 in a separate PR once that's working?

**Comment by overjt** [comment]
At this moment I am finishing the implementation, I already have the o1 models working, give me a few minutes

**Comment by notpeter** [comment]
Works for me. Have not reviewed the code.

**Comment by hmziqrs** [comment]
Hi, what's the priority on this PR? I had to move back to VS Code because of Claude. It's painful to work with VS Code copilot after experiencing Zed assistant.

**Comment by itsaphel** [comment]
Indeed, it'd be good to have a review on this soon! Excited to use Claude in Zed.

@notpeter @bennetbo

**Comment by bennetbo** [comment]
LGTM, tested both o1-mini and o1-preview. Claude 3.5 does not work for me, because I do not seem to have access just yet. Since it seems to work for others that do have access though, I'm going to merge this. Thanks!

File: crates/copilot/src/copilot_chat.rs [hunk 1]
```diff
@@ -68,6 +90,9 @@ impl Model {
             Self::Gpt4o => 128000,
             Self::Gpt4 => 8192,
             Self::Gpt3_5Turbo => 16385,
+            Self::O1Mini => 128000,
+            Self::O1Preview => 128000,
+            Self::Claude3_5Sonnet => 200_000,
```
Comment (humazed) [inline crates/copilot/src/copilot_chat.rs:95]: @overjt where these numbers are coming from?

Comment (overjt) [reply crates/copilot/src/copilot_chat.rs:95]: Those were the limits when I did the implementation (hours after GitHub's announcement), apparently the limits were changed in recent days.

The models and limits can be consulted in the copilot api
`https://api.individual.githubcopilot.com/models`

```json
{
  "data": [
    {
      "capabilities": {
        "family": "gpt-3.5-turbo",
        "limits": {
          "max_context_window_tokens": 16384,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 12288
        },
        "object": "model_capabilities",
        "supports": {
          "tool_calls": true
        },
        "tokenizer": "cl100k_base",
        "type": "chat"
      },
      "id": "gpt-3.5-turbo",
      "model_picker_enabled": false,
      "name": "GPT 3.5 Turbo",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-3.5-turbo-0613"
    },
    {
      "capabilities": {
        "family": "gpt-3.5-turbo",
        "limits": {
          "max_context_window_tokens": 16384,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 12288
        },
        "object": "model_capabilities",
        "supports": {
          "tool_calls": true
        },
        "tokenizer": "cl100k_base",
        "type": "chat"
      },
      "id": "gpt-3.5-turbo-0613",
      "model_picker_enabled": false,
      "name": "GPT 3.5 Turbo",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-3.5-turbo-0613"
    },
    {
      "capabilities": {
        "family": "gpt-4",
        "limits": {
          "max_context_window_tokens": 32768,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 32768
        },
        "object": "model_capabilities",
        "supports": {
          "tool_calls": true
        },
        "tokenizer": "cl100k_base",
        "type": "chat"
      },
      "id": "gpt-4",
      "model_picker_enabled": false,
      "name": "GPT 4",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4-0613"
    },
    {
      "capabilities": {
        "family": "gpt-4",
        "limits": {
          "max_context_window_tokens": 32768,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 32768
        },
        "object": "model_capabilities",
        "supports": {
          "tool_calls": true
        },
        "tokenizer": "cl100k_base",
        "type": "chat"
      },
      "id": "gpt-4-0613",
      "model_picker_enabled": false,
      "name": "GPT 4",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4-0613"
    },
    {
      "capabilities": {
        "family": "gpt-4o",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 64000
        },
        "object": "model_capabilities",
        "supports": {
          "parallel_tool_calls": true,
          "tool_calls": true
        },
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "gpt-4o",
      "model_picker_enabled": true,
      "name": "GPT 4o",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4o-2024-05-13"
    },
    {
      "capabilities": {
        "family": "gpt-4o",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 64000
        },
        "object": "model_capabilities",
        "supports": {
          "parallel_tool_calls": true,
          "tool_calls": true
        },
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "gpt-4o-2024-05-13",
      "model_picker_enabled": false,
      "name": "GPT 4o",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4o-2024-05-13"
    },
    {
      "capabilities": {
        "family": "gpt-4o",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 64000
        },
        "object": "model_capabilities",
        "supports": {
          "parallel_tool_calls": true,
          "tool_calls": true
        },
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "gpt-4-o-preview",
      "model_picker_enabled": false,
      "name": "GPT 4o",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4o-2024-05-13"
    },
    {
      "capabilities": {
        "family": "gpt-4o",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_output_tokens": 16384,
          "max_prompt_tokens": 64000
        },
        "object": "model_capabilities",
        "supports": {
          "parallel_tool_calls": true,
          "tool_calls": true
        },
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "gpt-4o-2024-08-06",
      "model_picker_enabled": false,
      "name": "GPT 4o",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4o-2024-08-06"
    },
    {
      "capabilities": {
        "family": "o1",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_prompt_tokens": 20000
        },
        "object": "model_capabilities",
        "supports": {},
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "o1-preview",
      "model_picker_enabled": true,
      "name": "o1-preview (Preview)",
      "object": "model",
      "preview": true,
      "vendor": "Azure OpenAI",
      "version": "o1-preview-2024-09-12"
    },
    {
      "capabilities": {
        "family": "o1",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_prompt_tokens": 20000
        },
        "object": "model_capabilities",
        "supports": {},
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "o1-preview-2024-09-12",
      "model_picker_enabled": false,
      "name": "o1-preview (Preview)",
      "object": "model",
      "preview": true,
      "vendor": "Azure OpenAI",
      "version": "o1-preview-2024-09-12"
    },
    {
      "capabilities": {
        "family": "o1-mini",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_prompt_tokens": 20000
        },
        "object": "model_capabilities",
        "supports": {},
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "o1-mini",
      "model_picker_enabled": true,
      "name": "o1-mini (Preview)",
      "object": "model",
      "preview": true,
      "vendor": "Azure OpenAI",
      "version": "o1-mini-2024-09-12"
    },
    {
      "capabilities": {
        "family": "o1-mini",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_prompt_tokens": 20000
        },
        "object": "model_capabilities",
        "supports": {},
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "o1-mini-2024-09-12",
      "model_picker_enabled": false,
      "name": "o1-mini (Preview)",
      "object": "model",
      "preview": true,
      "vendor": "Azure OpenAI",
      "version": "o1-mini-2024-09-12"
    },
    {
      "capabilities": {
        "family": "text-embedding-ada-002",
        "limits": {
          "max_inputs": 256
        },
        "object": "model_capabilities",
        "supports": {},
        "tokenizer": "cl100k_base",
        "type": "embeddings"
      },
      "id": "text-embedding-ada-002",
      "model_picker_enabled": false,
      "name": "Embedding V2 Ada",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "text-embedding-ada-002"
    },
    {
      "capabilities": {
        "family": "text-embedding-3-small",
        "limits": {
          "max_inputs": 512
        },
        "object": "model_capabilities",
        "supports": {
          "dimensions": true
        },
        "tokenizer": "cl100k_base",
        "type": "embeddings"
      },
      "id": "text-embedding-3-small",
      "model_picker_enabled": false,
      "name": "Embedding V3 small",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "text-embedding-3-small"
    },
    {
      "capabilities": {
        "family": "text-embedding-3-small",
        "object": "model_capabilities",
        "supports": {
          "dimensions": true
        },
        "tokenizer": "cl100k_base",
        "type": "embeddings"
      },
      "id": "text-embedding-3-small-inference",
      "model_picker_enabled": false,
      "name": "Embedding V3 small (Inference)",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "text-embedding-3-small"
    },
    {
      "capabilities": {
        "family": "gpt-4o-mini",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 12288
        },
        "object": "model_capabilities",
        "supports": {
          "parallel_tool_calls": true,
          "tool_calls": true
        },
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "gpt-4o-mini",
      "model_picker_enabled": false,
      "name": "GPT 4o Mini",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4o-mini-2024-07-18"
    },
    {
      "capabilities": {
        "family": "gpt-4o-mini",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 12288
        },
        "object": "model_capabilities",
        "supports": {
          "parallel_tool_calls": true,
          "tool_calls": true
        },
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "gpt-4o-mini-2024-07-18",
      "model_picker_enabled": false,
      "name": "GPT 4o Mini",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4o-mini-2024-07-18"
    },
    {
      "capabilities": {
        "family": "gpt-4-turbo",
        "limits": {
          "max_context_window_tokens": 128000,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 64000
        },
        "object": "model_capabilities",
        "supports": {
          "parallel_tool_calls": true,
          "tool_calls": true
        },
        "tokenizer": "cl100k_base",
        "type": "chat"
      },
      "id": "gpt-4-0125-preview",
      "model_picker_enabled": false,
      "name": "GPT 4 Turbo",
      "object": "model",
      "preview": false,
      "vendor": "Azure OpenAI",
      "version": "gpt-4-0125-preview"
    },
    {
      "capabilities": {
        "family": "claude-3.5-sonnet",
        "limits": {
          "max_context_window_tokens": 200000,
          "max_output_tokens": 4096,
          "max_prompt_tokens": 195000
        },
        "object": "model_capabilities",
        "supports": {},
        "tokenizer": "o200k_base",
        "type": "chat"
      },
      "id": "claude-3.5-sonnet",
      "model_picker_enabled": true,
      "name": "Claude 3.5 Sonnet (Preview)",
      "object": "model",
      "policy": {
        "state": "enabled",
        "terms": "Enable access to the latest Claude 3.5 Sonnet model from Anthropic. [Learn more about how GitHub Copilot serves Claude 3.5 Sonnet](https://docs.github.com/copilot/using-github-copilot/using-claude-sonnet-in-github-copilot)."
      },
      "preview": true,
      "vendor": "Anthropic",
      "version": "claude-3.5-sonnet"
    }
  ],
  "object": "list"
}```

**Comment by xcrap** [comment]
Do this update also supports the copilot custom instructions?

**Comment by notpeter** [comment]
> Do this update also supports the copilot custom instructions?

No. Please open a distinct issue for supporting `.github/copilot-instructions.md` files.

Zed does allow controlling the prompts used in the Assistant Panel (`assistant: deploy prompt library` from the command palette). 

Additionally, although there is no GUI, Zed also supports defining custom templates for other prompts by editing a few configuration files.
Screenshot 2024-11-08 at 12 32 47


See: https://zed.dev/docs/assistant/prompting#advanced-concepts

**Pull request merged**